### DISCUSSION TITLE
Prevent a refetch from triggering while a mutation is running?

### DISCUSSION THREAD

**Discussion by VanTanev** [TanStack/query#2245] [Q&A] [answered]
Given the following code:
```typescript
export const createMutation: UseMutationOptions<
  Todo,
  unknown,
  { title: string; completed?: boolean },
  { optimisticTodo: Todo }
> = {
  mutationKey: 'todos',
  mutationFn: create,
  onMutate: async (todo) => {
    await queryClient.cancelQueries("todos");
    const optimisticTodo: Todo = {
      id: uuid(),
      title: todo.title,
      completed: todo.completed ?? false,
    };
    queryClient.setQueryData<Todo[]>("todos", (todos = []) => [
      ...todos,
      optimisticTodo,
    ]);
    return { optimisticTodo };
  },
  onSuccess: (newTodo, _, ctx) => {
    queryClient.setQueryData<Todo[]>("todos", (todos = []) => {
      return todos.map((todo) =>
        todo.id === ctx?.optimisticTodo.id ? newTodo : todo
      );
    });
  },
  onError: (error, _, context) => {
    queryClient.setQueryData<Todo[]>("todos", (todos = []) =>
      todos.filter((todo) => todo.id !== context?.optimisticTodo?.id)
    );
  },
  onSettled: () => {
    queryCache.invalidateQueries('todos')
  }
};
```

I end up with the following behavior:

![](https://static.swimlanes.io/faeed6e085676b6269dc05877c4d8690.png)
https://swimlanes.io/u/atq5XoEmg

Is it possible to block any `todos` query from executing while a mutation is executing?

We can only say `cancelQueries()` at the start of the mutation, but that doesn't prevent another query from triggering while the mutation executes and overwriting the query cache with bad data.

**Comment by shamxeed** [comment]
Facing a similar issue...

**Comment by TkDodo** [comment]
I think the best way would be to disable your query if a mutation that could affect it is currently running. you can use the `useIsMutating` hook to find it, and then use the enabled option, something like:

```
const todosKey = ['todos']
const useTodos = () => {
  const isMutating = useIsMutating({ mutationKey: todosKey })
  return useQuery(todosKey, fetchTodos, { enabled: isMutating === 0 })
}
```

keys for `useIsMutating` are matched fuzzily, but you need to use a mutationKey for your mutations.

a more general but similar approach would be to disable queries globally if any mutation is running

**Reply by shamxeed** [reply]
Thank you so much it works... I was debugging this since morning, I also tried using the useIsMutating hook but my approach is wrong, thank you so much once again!!!

**Reply by VanTanev** [reply]
@TkDodo This feels incorrect to me. Not in the sense that it won't work, but in the sense that we're fixing a problem at the wrong abstraction level.

Given that react-query already has all this information, I believe it should automatically handle suspension of queries for ongoing mutations. If a mutation with a `queryKey` executes, all matching queries should suspend until the mutation completes. 
If a re-fetch was requested, it should execute after the mutation completes, and if no re-fetch was requested, no action should be taken.

Do you think this feature could make it into react-query proper instead of requiring fixes in userland?

**Reply by TkDodo** [reply]
> If a mutation with a queryKey executes

@VanTanev I see where you're getting at, but it's not that simple. react-query doesn't have an underlying schema, so mutations basically have nothing to do with queries. A mutation can never execute with a queryKey, it has a mutationKey. And because there is no schema, you have to do optimistic updates in user-land, and also do `invalidateQueries` manually. Because the lib can't know that updating a `todo` after a mutation needs to re-fetch `todos`, because, again, there is no schema (yet).

What I _can_ see us doing is providing an easier way to disable queries, so that you can disable them in `onMutate` and re-enable them in `onSettled` or so, instead of doing it directly on `useQuery`, so that it's a bit more explicit. Something like:

```diff
useMutation(updateTodo, {
   onMutate: async newTodo => {
     await queryClient.cancelQueries('todos')
+    queryClient.lockQueries('todos')    
 
     const previousTodos = queryClient.getQueryData('todos')
 
     queryClient.setQueryData('todos', old => [...old, newTodo])
 
     return { previousTodos }
   },
   onError: (err, newTodo, context) => {
     queryClient.setQueryData('todos', context.previousTodos)
   },
   onSettled: async () => {
+    queryClient.unlockQueries('todos')
     await queryClient.invalidateQueries('todos')  
   },
 })
```

I could even see _invalidateQueries_ automatically unlock them (maybe with an option), and `cancelQueries` automatically locking them (also maybe with an option). But I'm totally not sure here and would wait for other opinions ... @tannerlinsley ?

**Reply by VanTanev** [reply]
This sounds like a very reasonable approach in the short term. I hope we can see it added to react-query!

**Reply by shamxeed** [reply]
Sorry guys... But @TkDodo your last answer seemed a bit confusing to me as I can't find `lockQueries` method on `queryClient` hook, please where am I  going wrong?

**Reply by TkDodo** [reply]
It doesn’t exist. It was a proposal for the future :)

**Reply by shamxeed** [reply]
Okay thanks a lot yaarr!!!! 🙏🏻🙏🏻🙏🏻

**Reply by jzxchiang1** [reply]
Will this exist in v4?

**Reply by Steffen-Samseth** [reply]
I just tried this with react-query 3.39.3 and it doesn't seem to work. The original query is immediately invalidated and starts to refetch before the mutation starts, so useIsMutating returns 0. Is this something that has changed in later versions?

**Comment by jzxchiang1** [comment]
My solution based on this discussion was to add this to `useMutation`:

```
  onSettled: () => {
    // Only refetch if no other mutation is running
    if (queryClient.isMutating({mutationKey: todosKey}) === 1) {
      queryCache.invalidateQueries('todos')
    }
  }
  ```

**Comment by TkDodo** [comment, marked as answer]
> Will this exist in v4?

no, it won't. I don't think it would solve the underlying problem, which is:

- one mutation starts, locks queries
- a second mutations starts
- first mutation finishes, unlocks and refetches

unless the second mutations double-locks, we would still invalidate.

basically, we have been doing this in user-land lately:

```
export const useTrackParallelMutations = () => {
    const mutationNumber = React.useRef(0)

    return {
        startOne: () => {
            mutationNumber.current += 1
        },
        endOne: () => {
            if (mutationNumber.current > 0) {
                mutationNumber.current -= 1
            }
        },
        allEnded: () => mutationNumber.current === 0,
    }
}
```

with that, we can do:

```
const mutationTracker = useTrackParallelMutations()

useMutation(fn, {
  onMutate: () => mutationTracker.startOne()
  onSettled: () => {
    mutationTracker.endOne()
    if (mutationTracker.allEnded()) {
      queryClient.invalidateQueries(...)
    }
  }
})
```

> My solution based on this discussion was to add this to useMutation:

```
  onSettled: () => {
    // Only refetch if no other mutation is running
    if (queryClient.isMutating({mutationKey: todosKey}) === 1) {
      queryCache.invalidateQueries('todos')
    }
  }
```

this one also works, but you'd have to assign keys to all mutations. The `useTrackParallelMutations` hook only works for when the same custom hook is invoked many times (because it uses the same ref). If you trigger the same mutation from multiple places, checking `queryClient.isMutating` seems better 👍

**Reply by VanTanev** [reply]
This is a very good solution; Back in my use case, related mutations were in a single state machine, and I ended up tracking in-flight mutations in the state machine to achieve the same thing. It's great to have a React-native solution as well!
